File: replication/src/core/com/db4o/inside/replication/ReplicationProviderInside.java
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com */

package com.db4o.inside.replication;

import com.db4o.ext.Db4oUUID;
import com.db4o.foundation.Visitor4;
import com.db4o.replication.ReplicationProvider;


public interface ReplicationProviderInside extends ReplicationProvider {

	public void activate(Object object);

	public ReplicationReference referenceNewObject(Object obj, ReplicationReference counterpartReference, ReplicationReference referencingObjRef, String fieldName);

	/**
	 * Activates the fields, e.g. Collections, arrays, of an object
	 * <p/>
	 * /** Clear the  ReplicationReference cache
	 */
	void clearAllReferences();

	/**
	 * Destroys this provider and frees up resources.
	 */
	public void destroy();

	public void commitReplicationTransaction(long raisedDatabaseVersion);

	/**
	 * Returns the current transaction serial number.
	 *
	 * @return the current transaction serial number
	 */
	long getCurrentVersion();

	public Object getMonitor();

	ReadonlyReplicationProviderSignature getSignature();

	/**
	 * Determines if this Provider has the ReplicationReference of an object
	 *
	 * @param obj object concerned
	 * @return true if this provider has the ReplicationReference
	 */
	boolean hasReplicationReferenceAlready(Object obj);

	/**
	 * Returns the ReplicationReference of an object
	 *
	 * @param obj            object queried
	 * @param referencingObj
	 * @param fieldName
	 * @return null if the object is not owned by this ReplicationProvider.
	 */
	ReplicationReference produceReference(Object obj, Object referencingObj, String fieldName);

	/**
	 * Returns the ReplicationReference of an object by specifying the uuid of the object.
	 *
	 * @param uuid the uuid of the object
	 * @param hint the type of the object
	 * @return the ReplicationReference or null if the reference cannot be found
	 */
	ReplicationReference produceReferenceByUUID(Db4oUUID uuid, Class hint);

	boolean wasChangedSinceLastReplication(ReplicationReference reference);
	boolean wasDeletedSinceLastReplication(Db4oUUID uuid);


	/**
	 * Rollbacks all changes done during the replication session  and terminates the Transaction.
	 * Guarantees the changes will not be applied to the underlying databases.
	 */
	void rollbackReplication();

	/**
	 * Start a Replication Transaction with another ReplicationProvider
	 *
	 * @param peerSignature the signature of another ReplicationProvider.
	 */
	void startReplicationTransaction(ReadonlyReplicationProviderSignature peerSignature);


	/**
	 * Stores the new replicated state of obj. It can also be a new object to this
	 * provider.
	 *
	 * @param obj Object with updated state or a clone of new object in the peer.
	 */
	void storeReplica(Object obj);


	public void syncVersionWithPeer(long maxVersion);


	/**
	 * Visits the object of each cached ReplicationReference.
	 *
	 * @param visitor implements the visit functions, including copying of object states, and storing of changed objects
	 */
	void visitCachedReferences(Visitor4 visitor);

	String getName();
}

File: drs/drs.net/drs/src/com/db4o/drs/inside/ReplicationProviderInside.cs
namespace com.db4o.drs.inside
{
	public interface ReplicationProviderInside : com.db4o.drs.ReplicationProvider
	{
		void Activate(object _object);

		/// <summary>Activates the fields, e.g.</summary>
		/// <remarks>
		/// Activates the fields, e.g. Collections, arrays, of an object
		/// <p/>
		/// /** Clear the  ReplicationReference cache
		/// </remarks>
		void ClearAllReferences();

		void CommitReplicationTransaction(long raisedDatabaseVersion);

		/// <summary>Destroys this provider and frees up resources.</summary>
		/// <remarks>Destroys this provider and frees up resources.</remarks>
		void Destroy();

		com.db4o.ObjectSet GetStoredObjects(System.Type type);

		/// <summary>Returns the current transaction serial number.</summary>
		/// <remarks>Returns the current transaction serial number.</remarks>
		/// <returns>the current transaction serial number</returns>
		long GetCurrentVersion();

		long GetLastReplicationVersion();

		object GetMonitor();

		string GetName();

		com.db4o.drs.inside.ReadonlyReplicationProviderSignature GetSignature();

		/// <summary>Returns the ReplicationReference of an object</summary>
		/// <param name="obj">object queried</param>
		/// <param name="referencingObj"></param>
		/// <param name="fieldName"></param>
		/// <returns>null if the object is not owned by this ReplicationProvider.</returns>
		com.db4o.drs.inside.ReplicationReference ProduceReference(object obj, object referencingObj
			, string fieldName);

		/// <summary>Returns the ReplicationReference of an object by specifying the uuid of the object.
		/// 	</summary>
		/// <remarks>Returns the ReplicationReference of an object by specifying the uuid of the object.
		/// 	</remarks>
		/// <param name="uuid">the uuid of the object</param>
		/// <param name="hint">the type of the object</param>
		/// <returns>the ReplicationReference or null if the reference cannot be found</returns>
		com.db4o.drs.inside.ReplicationReference ProduceReferenceByUUID(com.db4o.ext.Db4oUUID
			 uuid, System.Type hint);

		com.db4o.drs.inside.ReplicationReference ReferenceNewObject(object obj, com.db4o.drs.inside.ReplicationReference
			 counterpartReference, com.db4o.drs.inside.ReplicationReference referencingObjRef
			, string fieldName);

		/// <summary>Rollbacks all changes done during the replication session  and terminates the Transaction.
		/// 	</summary>
		/// <remarks>
		/// Rollbacks all changes done during the replication session  and terminates the Transaction.
		/// Guarantees the changes will not be applied to the underlying databases.
		/// </remarks>
		void RollbackReplication();

		/// <summary>Start a Replication Transaction with another ReplicationProvider</summary>
		/// <param name="peerSignature">the signature of another ReplicationProvider.</param>
		void StartReplicationTransaction(com.db4o.drs.inside.ReadonlyReplicationProviderSignature
			 peerSignature);

		/// <summary>Stores the new replicated state of obj.</summary>
		/// <remarks>
		/// Stores the new replicated state of obj. It can also be a new object to this
		/// provider.
		/// </remarks>
		/// <param name="obj">Object with updated state or a clone of new object in the peer.
		/// 	</param>
		void StoreReplica(object obj);

		void SyncVersionWithPeer(long maxVersion);

		void UpdateCounterpart(object updated);

		/// <summary>Visits the object of each cached ReplicationReference.</summary>
		/// <remarks>Visits the object of each cached ReplicationReference.</remarks>
		/// <param name="visitor">implements the visit functions, including copying of object states, and storing of changed objects
		/// 	</param>
		void VisitCachedReferences(com.db4o.foundation.Visitor4 visitor);

		bool WasModifiedSinceLastReplication(com.db4o.drs.inside.ReplicationReference reference
			);

		void ReplicateDeletion(com.db4o.ext.Db4oUUID uuid);
	}
}
